lib/linked_list.rb

```diff
@@ -7,39 +7,103 @@ def initialize
     @head = nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def add_first(data)
-
+    if @head.nil? 
+      first_node=Node.new(data)
+      @head=first_node
+    else
+      first_node=Node.new(data,@head)
+      @head=first_node
+    end
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_first
-
+    if @head.nil?
+      return nil
+    else 
+      return @head.data
+    end 
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(n)
   def length
-    return 0
+    if @head.nil?
+       return 0
+    else 
+     return count_node(@head, 1) 
+    end
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
-  def add_last(data)
+  def count_node(node, counter)
+    if node.next == nil 
+      return counter 
+    else 
+      return count_node(node.next, counter += 1)
+    end
+  end
 
+  # Time complexity - O(n)
+  # Space complexity - O(n)
+  def add_last(data)
```

Comment:
The method works, but since you're not doing recursion the space complexity is O(1)